_posts/2019-01-28-from-tslint-to-eslint.md

```diff
@@ -0,0 +1,242 @@
+
+<!-- more -->
+
+This means that teams maintaining legacy JavaScript codebases will no longer have to _also_ maintain
+[two][tslint-react] [versions][eslint-react] of often nearly [identical][tslint-prettier]
+[rule-sets][eslint-prettier]. All of the aggregate culture that builds up around linting can now be shared in a
+forward and backward facing way, making the often-daunting process of migrating a codebase from JavaScript to
+TypeScript a much easier sell.
+
+With this in mind we wanted to give the new officially-sanctioned [typescript-eslint][typescript-eslint] project a
+spin and document our findings.
+
+### Setup
+
+To get started, install the necessary dependencies:
+
+```sh
+$ yarn install -D eslint typescript @typescript-eslint/eslint-plugin
+```
+
+Then create a new `.eslintrc.js` and add a bit of setup:
+
+```js
+module.exports = {
+  parser: "@typescript-eslint/parser",
+  plugins: ["@typescript-eslint"],
+  extends: ["eslint:recommended", "plugin:@typescript-eslint/recommended"],
+  parserOptions: {
+    ecmaVersion: 6,
+    project: "./tsconfig.json",
+    sourceType: "module"
+  }
+}
+```
+
+Note that `parserOptions.project` points to your `tsconfig.json` file:
+
+```json
+{
+  "compilerOptions": {}
+}
+```
+
+Next, add a bit of TypeScript to a file
+
+```sh
+$ echo "export const foo: any = 'bar'" > index.ts
+```
+
+and run the linter:
+
+```sh
+$ yarn eslint . --ext .ts,.tsx
+
+~/index.ts
+  1:12  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
+
+✖ 1 problem (0 errors, 1 warnings)
+```
+
+Very nice!
+
+Now lets expand the example a bit and add something more sophisticated, which in Artsy's use-case is commonly
+GraphQL:
+
+```sh
+$ yarn add -D eslint-plugin-graphql graphql-tag apollo
+```
+
+Update `tslint.json` and let it know we'll be using `node` for imports:
+
+```tslint.json
+{
+  "compilerOptions": {
+    "moduleResolution": "node"
+  }
+}
+```
+
+In `.eslintrc.js` add these rules (while noting the addition of `graphql` to `plugins` and
+`graphql/template-strings` under `rules`):
+
+```js
+const path = require("path")
+
+module.exports = {
+  parser: "@typescript-eslint/parser",
+  plugins: ["@typescript-eslint", "graphql"],
+  extends: ["eslint:recommended", "plugin:@typescript-eslint/recommended"],
+  parserOptions: {
+    ecmaVersion: 6,
+    project: "./tsconfig.json",
+    sourceType: "module"
+  },
+  rules: {
+    "graphql/template-strings": [
+      "error",
+      {
+        schemaJsonFilepath: path.resolve(__dirname, "./schema.json"),
+        tagName: "graphql"
+      }
+    ]
+  }
+}
+```
+
+For GraphQL to know what to lint, we'll need a schema. Thankfully the
+[Ethiopian Movie Database](https://etmdb.com/graphql) has our back :)
+
+```sh
+$ yarn apollo service:download --endpoint https://etmdb.com/graphql
```

Comment:
It was kinda a bummer not being able to easily download our schema due to the `__id` issue. Would be nice to somehow get that resolved so our environment is compatable with the larger ecosystem.

Reply:
The issue is probably that the tool tries to load the schema instance (without the ability to provide the allowed legacy names option), which doesn't seem necessary. A dependency-less alternative would be to do something like:

```
node -e 'const req = require("https").request("https://metaphysics-staging.artsy.net", { method: "POST", headers: { "content-type": "application/json" } }, res => { const chunks = []; res.on("data", c => chunks.push(c)); res.on("end", () => console.log(JSON.stringify(JSON.parse(Buffer.concat(chunks)).data, null, 2))); }); req.write(JSON.stringify({ query: require("graphql").getIntrospectionQuery() })); req.end();' > schema.json
```

…or a prettier version:

```
$ node -e 'const req = require("https").request(
  "https://metaphysics-staging.artsy.net",
  { method: "POST", headers: { "content-type": "application/json" } },
  res => {
    const chunks = [];
    res.on("data", c => chunks.push(c));
    res.on("end", () =>
      console.log(
        JSON.stringify(JSON.parse(Buffer.concat(chunks)).data, null, 2)
      )
    );
  }
);
req.write(
  JSON.stringify({ query: require("graphql").getIntrospectionQuery() })
);
req.end();
' > schema.json
```

Or, seeing as you're fine adding a dependency anyways, replace the `apollo` dep with `request` and run:

```
node -e 'require("request").post("https://metaphysics-staging.artsy.net", { json: { query: require("graphql").getIntrospectionQuery() } }, (_e, _r, { data }) => console.log(JSON.stringify(data, null, 2)))' > schema.json
```

Reply:
If it accepts a `.graphql` file then `curl https://github.com/artsy/metaphysics/raw/master/_schema.graphql` could work?

Reply:
Oh yeah, forgot we already dump it! 💥

Reply:
Since this is a blog post we should keep things concise and to the point so that the subject isn't lost. (For something internal that command would work fine, and could hide it behind a script.) But from a readers POV they shouldn't need to write code if the conventinal way to handle this is behind a single cli command. Most interested users will already haver already have `apollo` as a dependency anyways. 

I bet the `eslint` reads `.graphql` files, though... 👁. `curl` would work.

Reply:
I think this post is great and ready to go either way 👍

Reply:
Not sure what's up reading from our schema: 

<img width="1052" alt="screen shot 2019-01-29 at 9 18 15 am" src="https://user-images.githubusercontent.com/236943/51926778-ebda3400-23a6-11e9-88d9-58a67d4a140b.png">

If someone feels strongly about investigating / fixing we can iterate on it.

Reply:
I don’t feel strongly about using our schema, I was mostly responding to your comment about not being able to do so.